### PR TITLE
feat: dedupe shared helpers, expand test coverage, and harden persistence

### PR DESCRIPTION
**Pull request by jglopez**
Cleanup pass from a codebase-health review. Split into small commits so it's reviewable one fix at a time.

**Refactors (no behavior change)**
- Extract a shared percent/amount input parser; the same parse was reimplemented across 8 sites in \`calculate.ts\` and \`TaxTipSection\`.
- Share the paid-amount parser between settlement and the payer section; reuse the existing fee-base helper instead of a nested copy.
- Consolidate the currency formatters and drop the \`fmt\` aliases.
- Share the localStorage read/write try/catch between the versioned store and column-order hook.

**Tests**
- Add a shared test harness so the runner glue isn't copy-pasted across test files.
- Cover the previously-untested validators and formatters, plus breakdown/settlement edge cases (post-tax fees, zero-base splits, discounts, overpayment) and store/migration failure paths.
- Replace a tautological assertion in the settlement test.

**Robustness**
- Validate state restored from localStorage so a corrupt blob falls back instead of flowing into the math.
- Warn the user when changes can't be saved (private browsing / storage full) instead of failing silently.
- Cap participant, item, and fee name length.
- Drop an unsafe non-null assertion in item assignment.

**Tooling / docs**
- Scope the \`set-state-in-effect\` lint rule to its two legitimate uses instead of disabling it project-wide.
- Trim redundant and speculative comments.

Follow up: deciding whether to adopt Vitest + jsdom for hook/component coverage is tracked in #89.

### CODE REVIEW

File: src/lib/versionedStore.ts [hunk 1]
```diff
@@ -10,56 +10,67 @@ type Migration<T> = (raw: unknown) => T
...
 
         try {
-          const migrated = migrate(JSON.parse(raw))
-          localStorage.setItem(currentKey, JSON.stringify(migrated))
+          const migrated = migrate(raw)
+          writeJSON(currentKey, migrated)
           // Best-effort cleanup: a removeItem failure must not suppress the
           // migrated value that was already written to the current key.
           try { localStorage.removeItem(oldKey) } catch { /* ignore */ }
           return migrated
```
[the remark this answers is not in the document]
Comment (jglopez) [reply src/lib/versionedStore.ts:63]: Fixed in 1e341f3. The migration now gates the removeItem call on writeJSON's return value — if the write is blocked or quota-exceeded, the old key stays intact so the data survives the next load.

File: src/hooks/useBillSplit.ts [hunk 2]
```diff
@@ -20,10 +20,27 @@ const DEFAULT_STATE: BillState = {
...
+    Array.isArray(s.items) &&
+    typeof s.tax === 'string' &&
+    typeof s.tip === 'string' &&
+    (s.tipBase === 'pre-tax' || s.tipBase === 'post-tax') &&
+    Array.isArray(s.additionalFees) &&
+    (s.payerMode === 'single' || s.payerMode === 'multiple') &&
+    typeof s.singlePayerId === 'string' &&
+    typeof s.amountPaid === 'object' && s.amountPaid !== null && !Array.isArray(s.amountPaid)
+  )
+}
```
[the remark this answers is not in the document]
Comment (jglopez) [reply src/hooks/useBillSplit.ts]: Fixed in 6ffff61. isBillState now validates element shapes for all three arrays — isItem checks that price is a string, assignedTo is null or string[], and taxable is absent or boolean; isParticipant checks id/name strings; isAdditionalFee checks id/name/amount strings and a valid base literal. items:[null] and items:[{garbage}] now both fail validation and fall back to migration/default.

File: src/utils/calculate.ts [hunk 3]
```diff
@@ -132,6 +139,12 @@ export function getTotalFeeBase(base: FeesBase, totalSubtotal: number, totalTax:
   return base === 'post-tax' ? totalSubtotal + totalTax : totalSubtotal
 }
 
+/** Parse a paid-amount string, treating empty/null/undefined/NaN as 0. */
+export function parsePaidAmount(value: string | undefined): number {
+  const n = Number(value ?? '')
+  return isNaN(n) ? 0 : n
+}
```
[the remark this answers is not in the document]
Comment (jglopez) [reply src/utils/calculate.ts]: Fixed in 91baf3a. Swapped the isNaN guard for isFinite — Number("Infinity") is not NaN so the old guard let it through, which would poison settlement math. Also corrected the JSDoc; the type is string | undefined, null was never accepted.

File: src/utils/format.ts [hunk 4]
```diff
@@ -1,18 +1,16 @@
-/**
- * Format a number as a plain decimal amount (no $ prefix).
- * Negative amounts use the Unicode minus sign so meaning isn't conveyed
- * by color alone (WCAG). Used in summary table cells and item share hints.
- */
+// Negative amounts use Unicode minus (U+2212) so the sign precedes $ and
+// meaning isn't conveyed by color alone (WCAG).
```
[the remark this answers is not in the document]
Comment (jglopez) [reply src/utils/format.ts]: Fixed in 821fe57. Reworded to "precedes the currency symbol when present" — formatAmount has no $, so the original phrasing was only accurate for formatMoney.

File: src/utils/calculate.test.ts [hunk 5]
```diff
@@ -501,7 +482,209 @@ test('singlePayerId not in participants → no paid credit assigned', () => {
...
+    items: [item('x', '10')],
+    tax: '10%', // $1 tax, totalSubtotal = $10, post-tax = $11
+    additionalFees: [fee('svc', '10%', 'post-tax')], // 10% of $11 = $1.10
+  })
+  const bd = calculateBreakdown(s)
+  assertEqual(bd.totalTax, 1, '$1 tax')
+  assertEqual(bd.totalAdditionalFees[0], 1.1, '10% of $11 post-tax = $1.10')
+})
+
+test('percentage fee when pool base is 0 splits evenly', () => {
```
[the remark this answers is not in the document]
Comment (jglopez) [reply src/utils/calculate.test.ts]: Fixed in 1bb9504. The body tests the even-split fallback at zero pool base using a flat fee, not a percentage fee — renamed to "flat fee when pool base is 0 splits evenly".

File: src/lib/versionedStore.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jglopez) [reply src/lib/versionedStore.ts:63]: Note: Commit https://github.com/jglopez/bill-split/commit/1e341f36a743bd308932ee1102921479c6fb47fc was dropped. This was intentionally decided not to be fixed.
